src/agent/market_schedule.rs

```diff
@@ -0,0 +1,559 @@
+        // Prepare a schedule of narrow ranges
+        let market_schedule: MarketSchedule =
+            "UTC;O,O,O,O,O,O,O;0422/0900-1700,1109/0930-1730,1201/O,1225/C,1231/0930-2400"
+                .parse()
+                .unwrap();
+
+        let format = "%Y-%m-%d %H:%M";
+
+        // Date no match
+        assert!(market_schedule
+            .can_publish_at(&NaiveDateTime::parse_from_str("2023-11-20 05:30", format)?.and_utc()));
+
+        // Date match before range
+        assert!(!market_schedule
+            .can_publish_at(&NaiveDateTime::parse_from_str("2023-04-22 08:59", format)?.and_utc()));
+
+        // Date match at start of range
+        assert!(market_schedule
+            .can_publish_at(&NaiveDateTime::parse_from_str("2023-04-22 09:00", format)?.and_utc()));
+
+        // Date match in range
+        assert!(market_schedule
+            .can_publish_at(&NaiveDateTime::parse_from_str("2023-04-22 12:00", format)?.and_utc()));
+
+        // Date match at end of range
+        assert!(market_schedule
+            .can_publish_at(&NaiveDateTime::parse_from_str("2023-04-22 17:00", format)?.and_utc()));
+
+        // Date match after range
+        assert!(!market_schedule
+            .can_publish_at(&NaiveDateTime::parse_from_str("2023-04-22 17:01", format)?.and_utc()));
+
+        // Date 2400 range
+        assert!(market_schedule
+            .can_publish_at(&NaiveDateTime::parse_from_str("2023-12-31 23:59", format)?.and_utc()));
+        Ok(())
+    }
+}
+
+prop_compose! {
+    fn schedule_day_kind()(
+        r in any::<u8>(),
+        t1 in any::<u32>(),
+        t2 in any::<u32>(),
+    ) -> ScheduleDayKind {
+        match r % 3 {
+            0 => ScheduleDayKind::Open,
+            1 => ScheduleDayKind::Closed,
+            _ => ScheduleDayKind::TimeRange(
+                NaiveTime::from_hms_opt(t1 % 24, t1 / 24 % 60, 0).unwrap(),
+                NaiveTime::from_hms_opt(t2 % 24, t2 / 24 % 60, 0).unwrap(),
+            ),
+        }
+    }
+}
+
+prop_compose! {
+    fn holiday_day_schedule()(
+        m in 1..=12u32,
+        d in 1..=31u32,
+        s in schedule_day_kind(),
+    ) -> HolidayDaySchedule {
+        HolidayDaySchedule {
+            month: m,
+            day: d,
+            kind: s,
+        }
+    }
+}
+
+prop_compose! {
+    fn market_schedule()(
+        tz in proptest::sample::select(vec![
+            Tz::UTC,
+            Tz::America__New_York,
+            Tz::America__Los_Angeles,
+            Tz::America__Chicago,
+            Tz::Singapore,
+            Tz::Australia__Sydney,
+        ]),
+        weekly_schedule in proptest::collection::vec(schedule_day_kind(), 7..=7),
+        holidays in proptest::collection::vec(holiday_day_schedule(), 0..12),
+    ) -> MarketSchedule {
+        MarketSchedule {
+            timezone: tz,
+            weekly_schedule,
+            holidays,
+        }
+    }
+}
+
+const VALID_SCHEDULE_DAY_KIND_REGEX: &str =
```

Comment:
I recommend commenting what matches this regex.
It seems that it doesn't match hours > 12, is that intended?

Reply:
Thats an oversight. I fixed it to support 24 hour format.